File: api/src/main/java/org/glimpseframework/api/primitives/Vector.java
package org.glimpseframework.api.primitives;

import java.util.Objects;

/**
 * A vector in three-dimensional space, supporting affine transformations.
 *
 * <p>This class is immutable.</p>
 *
 * @author Slawomir Czerwinski
 */
public final class Vector {

	/**
	 * Creates a new vector.
	 * @param x coordinate x
	 * @param y coordinate y
	 * @param z coordinate z
	 */
	public Vector(float x, float y, float z) {
		this.x = x;
		this.y = y;
		this.z = z;
	}

	/**
	 * Creates a vector connecting two given points.
	 * @param start start point
	 * @param end end point
	 */
	public Vector(Point start, Point end) {
		this.x = end.getX() - start.getX();
		this.y = end.getY() - start.getY();
		this.z = end.getZ() - start.getZ();
	}

	/**
	 * Creates a vector from spherical coordinates.
	 * @param radius vector magnitude
	 * @param polarAngle vector polar angle
	 * @param azimuthAngle vector azimuth angle
	 * @return created vector
	 */
	public static Vector fromSphericalCoordinates(float radius, Angle polarAngle, Angle azimuthAngle) {
		return new Vector(
				Double.valueOf(radius * Math.sin(polarAngle.getRadians()) * Math.cos(azimuthAngle.getRadians()))
						.floatValue(),
				Double.valueOf(radius * Math.sin(polarAngle.getRadians()) * Math.sin(azimuthAngle.getRadians()))
						.floatValue(),
				Double.valueOf(radius * Math.cos(polarAngle.getRadians()))
						.floatValue());
	}

	/**
	 * Gets X coordinate of the vector.
	 * @return X coordinate
	 */
	public float getX() {
		return x;
	}

	/**
	 * Gets Y coordinate of the vector.
	 * @return Y coordinate
	 */
	public float getY() {
		return y;
	}

	/**
	 * Gets Z coordinate of the vector.
	 * @return Z coordinate
	 */
	public float getZ() {
		return z;
	}

	/**
	 * Gets magnitude of the vector.
	 * @return vector's magnitude
	 */
	public float getMagnitude() {
		return Double.valueOf(Math.sqrt(x * x + y * y + z * z)).floatValue();
	}

	/**
	 * Gets polar angle of the vector.
	 * @return vector's polar angle
	 */
	public Angle getPolarAngle() {
		return Angle.fromRadians(Double.valueOf(Math.atan2(Math.sqrt(x * x + y * y), z)).floatValue());
	}

	/**
	 * Gets azimuth angle of the vector.
	 * @return vector's azimuth angle
	 */
	public Angle getAzimuthAngle() {
		return Angle.fromRadians(Double.valueOf(Math.atan2(y, x)).floatValue());
	}

	/**
	 * Checks whether the vector is parallel to another vector.
	 * @param v second vector
	 * @return {@code true} if the vectors are parallel; otherwise returns {@code false}
	 */
	public boolean isParallelTo(Vector v) {
		return (x * v.y == v.x * y) && (x * v.z == v.x * z) && (y * v.z == v.y * z);
	}

	/**
	 * Gets a vector with the same direction, but with a magnitude of 1.
	 * @return vector with a magnitude of 1
	 * @throws NullVectorNormalizationException if vector's magnitude is 0.
	 */
	public Vector normalize() throws NullVectorNormalizationException {
		return normalize(1.0f);
	}

	/**
	 * Gets a vector with the same direction, but with a given magnitude.
	 * <p>If the {@code magnitude} argument is negative,
	 * the direction of the resulting vector will be changed to opposite.</p>
	 * @param magnitude magnitude of the new vector
	 * @return vector with the given magnitude
	 * @throws NullVectorNormalizationException if vector's magnitude is 0.
	 */
	public Vector normalize(float magnitude) throws NullVectorNormalizationException {
		if (equals(NULL_VECTOR)) {
			throw new NullVectorNormalizationException();
		}
		return multiply(magnitude / getMagnitude());
	}

	/**
	 * Creates a new vector, being a result of scalar multiplication.
	 * <p>The new vector has the same direction as the original vector, but its magnitude is multiplied
	 * by {@code scalar}.</p>
	 * <p>If the {@code scalar} argument is negative, the direction will be changed to opposite.</p>
	 * @param scalar multiplication factor
	 * @return vector with the multiplied magnitude
	 */
	public Vector multiply(float scalar) {
		return new Vector(scalar * x, scalar * y, scalar * z);
	}

	/**
	 * Creates a vector being a cross product of two vectors.
	 * @param v second vector
	 * @return cross product of two vectors
	 */
	public Vector crossProduct(Vector v) {
		return new Vector(
				y * v.z - z * v.y,
				z * v.x - x * v.z,
				x * v.y - y * v.x);
	}

	/**
	 * Gets vector coordinates as an array of 3 float numbers.
	 * @return array of 3 float numbers
	 */
	public float[] get3f() {
		return new float[] {x, y, z};
	}

	/**
	 * Gets vector coordinates in augmented form, as an array of 4 float numbers.
	 * <p>4th number is always 1.0f.</p>
	 * @return array of 4 float numbers
	 */
	public float[] get4f() {
		return new float[] {x, y, z, 1.0f};
	}

	@Override
	public String toString() {
		return String.format("[%.1f %.1f %.1f 1.0]", x, y, z);
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) {
			return true;
		}
		if (o == null || getClass() != o.getClass()) {
			return false;
		}

		Vector vector = (Vector) o;

		if (Float.compare(vector.x, x) != 0) {
			return false;
		}
		if (Float.compare(vector.y, y) != 0) {
			return false;
		}
		if (Float.compare(vector.z, z) != 0) {
			return false;
		}

		return true;
	}

	@Override
	public int hashCode() {
		return Objects.hash(x, y, z);
	}

	/** A null vector. */
	public static final Vector NULL_VECTOR = new Vector(0.0f, 0.0f, 0.0f);

	/** A unit vector parallel to X-axis. */
	public static final Vector X_UNIT_VECTOR = new Vector(1.0f, 0.0f, 0.0f);

	/** A unit vector parallel to Y-axis. */
	public static final Vector Y_UNIT_VECTOR = new Vector(0.0f, 1.0f, 0.0f);

	/** A unit vector parallel to Z-axis. */
	public static final Vector Z_UNIT_VECTOR = new Vector(0.0f, 0.0f, 1.0f);

	private final float x, y, z;
}

File: api/src/main/kotlin/glimpse/api/Vector.kt
package glimpse.api

/**
 * A vector in three-dimensional space.
 *
 * @param x X coordinate.
 * @param y Y coordinate.
 * @param z Z coordinate.
 */
data class Vector(val x: Float, val y: Float, val z: Float) {

	/**
	 * Constructs a [Vector] from spherical coordinates.
	 *
	 * @param radius Radial coordinate.
	 * @param polarAngle Polar angle.
	 * @param azimuthAngle Azimuth angle.
	 */
	constructor(radius: Float, polarAngle: Angle, azimuthAngle: Angle) : this(
			radius * sin(polarAngle) * cos(azimuthAngle),
			radius * sin(polarAngle) * sin(azimuthAngle),
			radius * cos(polarAngle))

	companion object {

		/**
		 * Null vector.
		 */
		val NULL = Vector(0f, 0f, 0f)

		/**
		 * Unit vector in the direction of the X axis.
		 */
		val X_UNIT = Vector(1f, 0f, 0f)

		/**
		 * Unit vector in the direction of the Y axis.
		 */
		val Y_UNIT = Vector(0f, 1f, 0f)

		/**
		 * Unit vector in the direction of the Z axis.
		 */
		val Z_UNIT = Vector(0f, 0f, 1f)
	}

	internal val _3f: Array<Float> by lazy { arrayOf(x, y, z) }
	internal val _4f: Array<Float> by lazy { arrayOf(x, y, z, 1f) }

	/**
	 * Magnitude of the [Vector].
	 */
	val magnitude by lazy { Math.sqrt((x * x + y * y + z * z).toDouble()).toFloat() }

	/**
	 * Polar angle of the vector in spherical coordinate system.
	 */
	val polarAngle by lazy { atan2(Vector(x, y, 0f).magnitude, z) }

	/**
	 * Azimuth angle of the vector in spherical coordinate system.
	 */
	val azimuthAngle by lazy { atan2(y, x) }

	/**
	 * Returns a unit vector in the direction of this [Vector].
	 */
	val normalize by lazy { normalize(1f) }

	/**
	 * Returns a vector opposit to this vector.
	 */
	operator fun unaryMinus() = Vector(-x, -y, -z)

	/**
	 * Returns a sum of this vector and the [other] vector.
	 */
	operator fun plus(other: Vector) = Vector(x + other.x, y + other.y, z + other.z)

	/**
	 * Returns a difference of this vector and the [other] vector.
	 */
	operator fun minus(other: Vector) = Vector(x - other.x, y - other.y, z - other.z)

	/**
	 * Returns a cross product of this vector and the [other] vector.
	 */
	operator fun times(other: Vector): Vector = Vector(
			y * other.z - z * other.y,
			z * other.x - x * other.z,
			x * other.y - y * other.x)

	/**
	 * Returns a dot product of this vector and the [other] vector.
	 */
	infix fun dot(other: Vector): Float = throw UnsupportedOperationException()

	/**
	 * Returns a product of this vector and the [number].
	 */
	operator fun times(number: Float) = Vector(x * number, y * number, z * number)

	/**
	 * Returns a quotient of this vector and the [number].
	 */
	operator fun div(number: Float) = Vector(x / number, y / number, z / number)

	/**
	 * Returns `true` if this vector is parallel to the [other] vector.
	 */
	infix fun parallelTo(other: Vector) = x * other.y == other.x * y && x * other.z == other.x * z && y * other.z == other.y * z

	/**
	 * Returns a vector of the given [magnitude], in the direction of this vector.
	 */
	fun normalize(magnitude: Float): Vector {
		require(magnitude > 0f) { "Cannot normalize a null vector. Division by 0." }
		return this * magnitude / this.magnitude
	}

	/**
	 * Converts the [Vector] to a [Point].
	 */
	fun toPoint() = Point(x, y, z)
}